pkg/controller/deployment_util.go

```diff
@@ -390,6 +390,56 @@ func SetNewMachineSetNodeTemplate(deployment *v1alpha1.MachineDeployment, newIS
 	return nodeTemplateChanged
 }
 
+// SetNewMachineSetConfig sets new machine set's config appropriately by updating its revision and
+// copying required deployment nodeTemplates to it; it returns true if machine set's config is changed.
+func SetNewMachineSetConfig(deployment *v1alpha1.MachineDeployment, newIS *v1alpha1.MachineSet, newRevision string, exists bool) bool {
+	// First, copy deployment's config
+	configChanged := copyMachineDeploymentConfigToMachineSet(deployment, newIS)
+	// Then, update machine set's revision annotation
+	if newIS.Annotations == nil {
+		newIS.Annotations = make(map[string]string)
+	}
+	oldRevision, ok := newIS.Annotations[RevisionAnnotation]
+
+	oldRevisionInt, err := strconv.ParseInt(oldRevision, 10, 64)
+	if err != nil {
+		if oldRevision != "" {
+			klog.Warningf("Updating machine set revision OldRevision not int %s", err)
+			return false
+		}
+		//If the RS annotation is empty then initialise it to 0
+		oldRevisionInt = 0
+	}
+	newRevisionInt, err := strconv.ParseInt(newRevision, 10, 64)
+	if err != nil {
+		klog.Warningf("Updating machine set revision NewRevision not int %s", err)
+		return false
+	}
+	if oldRevisionInt < newRevisionInt {
+		newIS.Annotations[RevisionAnnotation] = newRevision
+		configChanged = true
+		klog.V(4).Infof("Updating machine set %q revision to %s", newIS.Name, newRevision)
+	}
+	// If a revision annotation already existed and this machine set was updated with a new revision
+	// then that means we are rolling back to this machine set. We need to preserve the old revisions
+	// for historical information.
+	if ok && configChanged {
+		revisionHistoryAnnotation := newIS.Annotations[RevisionHistoryAnnotation]
+		oldRevisions := strings.Split(revisionHistoryAnnotation, ",")
+		if len(oldRevisions[0]) == 0 {
+			newIS.Annotations[RevisionHistoryAnnotation] = oldRevision
+		} else {
+			oldRevisions = append(oldRevisions, oldRevision)
+			newIS.Annotations[RevisionHistoryAnnotation] = strings.Join(oldRevisions, ",")
+		}
+	}
+	// If the new machine set is about to be created, we need to add replica annotations to it.
+	if !exists && SetReplicasAnnotations(newIS, (deployment.Spec.Replicas), (deployment.Spec.Replicas)+MaxSurge(*deployment)) {
+		configChanged = true
+	}
```

Comment:
I have simply adopted it from the other 2 parallel methods used in the same check: `SetNewMachineSetAnnotations`, `SetNewMachineSetNodeTemplates`, any changes to these 2 methods could otherwise break the flow.

Reply:
I have tested this flow in the OOT migration. I think a lot of stuff is redundant here as this logic keeps running multiple times. This is some bare minimum that is required i think - https://github.com/prashanth26/machine-controller-manager/commit/5be393e3ea43f59d3b894c97f4389b53fb4472d5#diff-3e28f59dda362e1974d0f3df011e28f1R393. However, if you aren't sure. We can keep this for now, however, we will need to revist it in the future.

Reply:
Sure, I think there is some redundancy in couple of methods, which can be optimized with a bit of refactoring, but outside the scope of this PR.

Reply:
one side effect would be that if all - `SetNewMachineSetAnnotations()`, `SetNewMachineSetNodeTemplate()` and now `SetNewMachineSetConfig()` end up changing something, then we would have triplicates of `revisionHistoryAnnotation` for each revision in the machine deployment, because they all `append` it.

Reply:
Yes, we need to solve this more holistically, by carving the common code/logic out. We should probably open an issue to track this optimization separately.

Reply:
sure. should be fine